src/ledctl/ledctl.c

```diff
@@ -555,73 +779,109 @@ static led_status_t _cmdline_ibpi_parse(int argc, char *argv[])
 static led_status_t _cmdline_parse_non_root(int argc, char *argv[])
 {
 	int opt_index, opt = -1;
-	led_status_t status = LED_STATUS_SUCCESS;
+	char *shortopts;
+	struct option *longopts;
 
-	do {
-		opt = getopt_long(argc, argv, shortopt, longopt, &opt_index);
-		switch (opt) {
-		case 'v':
+	optind = test_params ? optind : 1;
+
+	setup_options(&longopts, &shortopts, possible_params, ARRAY_SIZE(possible_params));
+
+	opt = getopt_long(argc, argv, shortopts, longopts, &opt_index);
+	switch (opt) {
+	case 'v':
+		if (!test_params)
 			_ledctl_version();
-			exit(EXIT_SUCCESS);
-		case 'h':
+		exit(EXIT_SUCCESS);
+	case 'h':
+		if (!test_params)
 			_ledctl_help();
-			exit(EXIT_SUCCESS);
-		case ':':
-		case '?':
-			return LED_STATUS_CMDLINE_ERROR;
-		}
-	} while (opt >= 0);
+		exit(EXIT_SUCCESS);
+	case ':':
+	case '?':
+		free(shortopts);
+		free(longopts);
+		return LED_STATUS_CMDLINE_ERROR;
+#ifdef ENABLE_TEST
+	case 'T':
+		test_params = true;
+		return _cmdline_parse_non_root(argc, argv);
+#endif
+	}
 
-	return status;
+	free(shortopts);
+	free(longopts);
+
+	return LED_STATUS_SUCCESS;
 }
 
-/**
- * @brief Inits slot request structure with initial values.
- *
- * @param[in]       slot_req       structure with slot request
- *
- * @return This function does not return a value.
- */
-static void slot_request_init(struct slot_request *slot_req)
+led_status_t _cmdline_validate_non_option_args(int argc, char *argv[], struct request *req)
 {
-	memset(slot_req, 0, sizeof(struct slot_request));
+	int opt_index = -1;
+	char *shortopts;
+	struct option *longopts;
+	bool exit = false;
+
+	optind = 1;
+	setup_options(&longopts, &shortopts, possible_params, ARRAY_SIZE(possible_params));
+
+	// parse all arguments, invalid ones will be placed on the end of array argv[]
```

Comment:
bad comment style

Reply:
If someone will mess it up like:
`ledctl --listed-only locate={ /dev/sdb } -x rebuild=/sys/block/sdc`
it is not longer true valid, or it will be parsed?

Cannot it be merged or executed in cmdline_parse?

Reply:
It will be parsed and according to man it should be supported.